larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewThreeDKinkTool.cc

```diff
@@ -0,0 +1,451 @@
+/**
+ *  @file   larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewThreeDKinkTool.cc
+ *
+ *  @brief  Implementation of the two view three d kink tool class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+#include "larpandoracontent/LArHelpers/LArGeometryHelper.h"
+#include "larpandoracontent/LArHelpers/LArPointingClusterHelper.h"
+#include "larpandoracontent/LArObjects/LArPointingCluster.h"
+
+#include "larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewThreeDKinkTool.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+TwoViewThreeDKinkTool::TwoViewThreeDKinkTool() :
+    m_minXOverlapFraction(0.1f),
+    m_minMatchingScore(0.95f),
+    m_minLocallyMatchedFraction(0.3f),
+    m_minLongitudinalImpactParameter(-1.f),
+    m_nLayersForKinkSearch(10),
+    m_additionalXStepForKinkSearch(0.01f),
+    m_maxTransverseImpactParameter(5.f),
+    m_minImpactParameterCosTheta(0.5f),
+    m_cosThetaCutForKinkSearch(0.75f)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TwoViewThreeDKinkTool::~TwoViewThreeDKinkTool()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewThreeDKinkTool::Run(TwoViewTransverseTracksAlgorithm *const pAlgorithm, MatrixType &overlapMatrix)
+{
+    if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
+        std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    ModificationList modificationList;
+    this->GetModifications(pAlgorithm, overlapMatrix, modificationList);
+    const bool changesMade(this->ApplyChanges(pAlgorithm, modificationList));
+
+    return changesMade;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewThreeDKinkTool::PassesElementCuts(MatrixType::ElementList::const_iterator eIter, const ClusterSet &usedClusters) const
+{
+    if (usedClusters.count(eIter->GetCluster1()) || usedClusters.count(eIter->GetCluster2()))
+        return false;
+
+    if (eIter->GetOverlapResult().GetTwoViewXOverlap().GetXOverlapFraction0() - m_minXOverlapFraction < -std::numeric_limits<float>::epsilon())
+        return false;
+
+    if (eIter->GetOverlapResult().GetTwoViewXOverlap().GetXOverlapFraction1() - m_minXOverlapFraction < -std::numeric_limits<float>::epsilon())
```

Comment:
How come these two use `-std::numeric_limits<float>::epsilon()` and the other checks drop the minus sign?

Reply:
Ah, these cuts are the same as those implemented in the TwoViewClearTracksTool. Not sure why. I would lean to adding a minus sign to the following two comparisons too, what do you reckon?

Reply:
No strong preference from me, although would make sense for the two tools mentioned above to be consistent.

(Only a small snippet, so can live with separate implementations for now, unless there's a _very_ simple way to share implementation.)

Reply:
probably means I included the minus sign then.  Honestly no idea why...

Reply:
It was most likely me, @absolution1!
As for @johnmarshall80's suggestion to share implementation, could maybe think of adding a PassesElementCuts method to TransverseMatrixTool, that some (but not all) of the derived tools would make use of?

Reply:
I'd advocate for not tying the two tools together in this way.  I can imagine that we may want to update the implementations in a non-uniform way in the future and having them both be tied to each other just makes that procedure more difficult.
If there were huge chunks of the tools that were boilerplated then, sure, but we're talking about single functions that contain ~5 if statements, right?